### PR TITLE
[GEN] Enhance matrix operators verification

### PR DESCRIPTION
**Pull request by whitneywhtsang**

### CODE REVIEW

File: third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp [hunk 1]
```diff
@@ -173,21 +197,42 @@ LogicalResult TritonGEN::MatrixDPASOp::verify() {
...
+  if (verifyMatrixInput(*this).failed())
+    return failure();
+
+  VectorType resTy = getRes().getType();
+  unsigned resSize =
+      resTy.getNumElements() * resTy.getElementType().getIntOrFloatBitWidth();
+  unsigned subgroupSize = TritonGEN::getSubgroupSize(*this);
+  if (resSize * subgroupSize !=
+      getElemSizeInBits() * getTileHeight() * getTileWidth() * getVBlocks())
+    return emitOpError("result size does not match the expected size");
```
Comment (etiotto) [inline third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp:209]: can the expected value be printed in the msg?

File: third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp [hunk 2]
```diff
@@ -36,43 +42,61 @@ template <typename Op> static LogicalResult verifyInput(Op op) {
...
 
+template <typename Op> static LogicalResult verifyMatrixReadInput(Op op) {
+  static_assert(llvm::is_one_of<Op, TritonGEN::Matrix2DBlockLoadOp,
+                                TritonGEN::Matrix2DBlockPrefetchOp>::value,
+                "Unexpected template parameter");
+
+  uint32_t tileWidth = op.getTileWidth();
+  if (op.getVnniTransform()) {
+    if (tileWidth != 16)
+      return op->emitOpError("tile_width for vnni transform should be equal "
```
Comment (etiotto) [inline third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp:86]: tile_width when vnni_transform is true should ...

File: third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp [hunk 3]
```diff
@@ -36,43 +42,61 @@ template <typename Op> static LogicalResult verifyInput(Op op) {
...
+                             "to subgroup size, i.e., 16 elements");
+    return success();
+  }
+
+  // When reading matrix B of 32 bit elements, it does not need to be vnni transformed.
+  if (op.getElemSizeInBits() == 32 && tileWidth != 8 && tileWidth != 16)
+    return op->emitOpError("tile_width for 32 bit elements should be equal "
```
Comment (etiotto) [inline third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp:93]: `tile_width for 32 bit elements should be equal to systolic depth (8 elements) for matrix A and the subgroup size (print value of subgroup size) for matrix B`

File: third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp [hunk 4]
```diff
@@ -36,43 +42,61 @@ template <typename Op> static LogicalResult verifyInput(Op op) {
...
-  std::optional<int64_t> width = getConstantIntValue(op.getBaseWidth());
-  std::optional<int64_t> pitch = getConstantIntValue(op.getBasePitch());
-  if (pitch && width && *pitch < *width)
-    return op->emitOpError(
-        "4th operand (base pitch) should be >= 2nd operand (base width)");
+  if (op.getTranspose() && op.getElemSizeInBits() != 32)
+    return op->emitOpError("transpose is only supported for 32 bit elements");
+
+  if (op.getVnniTransform() && op.getElemSizeInBits() == 32)
+    return op->emitOpError("vnni transform is only supported for 8 and 16 bit "
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp]: done.

File: third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp (hunk 2 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp]: done.

File: third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp (hunk 3 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp]: done.

File: third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/Dialect/TritonGEN/IR/TritonGENOps.cpp]: done.

**Review by etiotto — APPROVED** [review]
Just left some inline comments

File: test/TritonGEN/tritongen-invalid.mlir [hunk 5]
```diff
@@ -177,133 +177,197 @@ llvm.func @triton_gen.dpas(%c : vector<8xf32>, %a : vector<8xi16>, %b : vector<8
...
+  %0 = triton_gen.2Dblockload %ptr, %base_width, %base_height, %base_pitch, %x, %y {elem_size_in_bits=8, tile_width=16, tile_height=8, v_blocks=1, transpose=false, vnni_transform=false, cache_control=Default} : (!llvm.ptr, i32, i32, i32, i32, i32) -> vector<4xi16>
   llvm.return
 }
 
 // -----
 
 llvm.func @matrix_2Dblockload(%ptr : !llvm.ptr, %base_width : i32, %base_height : i32, %base_pitch : i32, %x : i32, %y : i32) {
-  // expected-error @+1 {{'triton_gen.2Dblockload' op expecting tile_height to be 1, 2, 4, 8, 16, or 32}}
-  %0 = triton_gen.2Dblockload %ptr, %base_width, %base_height, %base_pitch, %x, %y {elem_size_in_bits=32, tile_width=8, tile_height=64, v_blocks=1, transpose=false, vnni_transform=false, cache_control=Default} : (!llvm.ptr, i32, i32, i32, i32, i32) -> vector<8xf32>
+  // expected-error @+1 {{'triton_gen.2Dblockload' op tile_width for 32 bit elements should be equal to systolic depth (8 elements) for matrix A and the subgroup size (16 elements) for matrix B}}
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply test/TritonGEN/tritongen-invalid.mlir]: done.

**Pull request merged**